apps/passport-client/components/modals/AccountImportModal.tsx

```diff
@@ -0,0 +1,253 @@
+            return false;
+          }
+
+          // If a PCD with this ID exists already, don't import it
+          if (pcdCollection.hasPCDWithId(pcd.id)) {
+            return false;
+          }
+
+          // Otherwise, do import it
+          return true;
+        };
+
+        const pcdsToMerge = importedCollection.getAll().filter(preImportFilter);
+        const pcdsToMergeIds = new Set(pcdsToMerge.map((pcd) => pcd.id));
+
+        // Because this hook can be called multiple times, we should only
+        // update the state if something has really changed
+        if (
+          // If the previous state was "ready", there's definitely a change
+          importState.state === "ready" ||
+          // If we have a new imported collection object, that's a change
+          importState.collection !== importedCollection ||
+          // If the set of PCDs to merge is different, that's a change
+          importState.pcdsToMergeIds.size !== pcdsToMergeIds.size ||
+          [...pcdsToMergeIds].find((id) => !importState.pcdsToMergeIds.has(id))
+        ) {
+          setImportState({
+            state: "valid-file-selected",
+            collection: importedCollection,
+            pcdsToMergeIds: new Set(pcdsToMerge.map((pcd) => pcd.id))
+          });
+        }
+      }
+    })();
+  }, [filesContent, importState, pcdCollection]);
+
+  return (
+    <Container>
+      <Spacer h={24} />
+      {importState.state === "ready" && (
+        <>
+          <p>
+            If you have previously exported a backup of your account, you can
+            restore any lost PCDs by importing the backup data.
+          </p>
+          <p>Importing data will not overwrite any of your existing PCDs.</p>
+          <p>To begin, select a backup file by clicking the button below.</p>
+          <Spacer h={8} />
+          <Button onClick={() => openFilePicker()}>Select file</Button>
+          <Spacer h={8} />
+        </>
+      )}
+      {importState.state === "valid-file-selected" && (
+        <>
+          {importState.pcdsToMergeIds.size == 0 && (
+            <>
+              <p>
+                The selected file does not contain any new PCDs. You may try to
+                restore from another backup.
+              </p>
+              <Spacer h={8} />
+              <Button onClick={() => openFilePicker()}>Select file</Button>
+              <Spacer h={8} />
+            </>
+          )}
+          {importState.pcdsToMergeIds.size > 0 && (
+            <>
+              <p>
+                The selected file contains{" "}
+                <strong>{importState.pcdsToMergeIds.size}</strong> new PCDs.
+              </p>
+              <Spacer h={8} />
+              <Button onClick={importPCDs}>Import PCDs</Button>
+              <Spacer h={8} />
+            </>
+          )}
+        </>
+      )}
+      {importState.state === "import-complete" &&
+        modal.modalType === "account-import" &&
+        !modal.error && (
```

Comment:
this is the sort of thing that makes modal-based UI implementations somewhat confusing for me to follow. #1346.

Reply:
What part is hard to follow? The alternative is adding new top-level `AppState` fields (and selectors for them, and so on), and we'd be checking those values here instead. It would eliminate the first check on `modalType` for type-narrowing purposes - is that the problem?